nntrainer/layers/cl_layers/concat_cl.cpp

```diff
@@ -233,12 +233,54 @@ std::string concat_cl_axis1_kernel_ =
 })";
 
 namespace nntrainer {
-ConcatLayerCl::ConcatLayerCl() : Layer() {}
+ConcatLayerCl::ConcatLayerCl() : LayerImplCl() {}
 
 static constexpr size_t SINGLE_INOUT_IDX = 0;
 static constexpr size_t INPUT_IDX_1 = 0;
 static constexpr size_t INPUT_IDX_2 = 1;
 
+bool ConcatLayerCl::registerClKernels() {
+  ClContext::SharedPtrClKernel kernel_concat_ptr = nullptr;
+
+  kernel_concat_ptr =
+    cl_context_ref.registerClKernel(concat_cl_axis1_kernel_, "concat_cl_axis1");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis1 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr =
+    cl_context_ref.registerClKernel(concat_cl_axis2_kernel_, "concat_cl_axis2");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis2 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr =
+    cl_context_ref.registerClKernel(concat_cl_axis3_kernel_, "concat_cl_axis3");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis3 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr = cl_context_ref.registerClKernel(
+    concat_cl_axis1_kernel_fp16_, "concat_cl_axis1_fp16");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis1_fp16 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr = cl_context_ref.registerClKernel(
+    concat_cl_axis2_kernel_fp16_, "concat_cl_axis2_fp16");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis2_fp16 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
+
+  kernel_concat_ptr = cl_context_ref.registerClKernel(
+    concat_cl_axis3_kernel_fp16_, "concat_cl_axis3_fp16");
+  NNTR_THROW_IF(!kernel_concat_ptr, std::runtime_error)
+    << "OpenCL Error: Fail to register concat_cl_axis3_fp16 kernel";
+  layer_kernel_ptrs.emplace_back(kernel_concat_ptr);
```

Comment:
wouldn't this throw an error when fp16 is not enabled?

Reply:
Yes you're right. I just write the code based on the current kernel version. I will add a separate commit to update this. Thanks!